src/Microsoft.AspNet.Server.Kestrel/Http/SocketOutput.cs

```diff
@@ -146,8 +151,16 @@ public class SocketOutput : ISocketOutput
                 else
                 {
                     // immediate write, which is not eligable for instant completion above
-                    tcs = new TaskCompletionSource<object>(buffer.Count);
-                    _tasksPending.Enqueue(tcs);
+                    tcs = new TaskCompletionSource<object>();
+                    if (!cancellationToken.IsCancellationRequested)
+                    {
+                        _tasksPending.Enqueue(new WaitingTask() { Token = cancellationToken, BytesToWrite = buffer.Count, CompletionSource = tcs });
```

Comment:
I think we are going to need to register with the `cancellationToken` if it [CanBeCanceled](https://msdn.microsoft.com/en-us/library/system.threading.cancellationtoken.canbecanceled(v=vs.110).aspx). If the callback is triggered before the write completes we should abort the connection.